### DISCUSSION TITLE
fireflux lnd->atm summary (with application to FATES and oslo_aero)

### DISCUSSION THREAD

**Discussion by mvertens** [NorESMhub/CTSM#46] [General]
Starting from the following

**drv_flds_in:**
------------------

```
&fire_emis_nl
  fire_emis_elevated = .false.
  fire_emis_factors_file = '/glade/campaign/cgd/tss/lawrence/ccisensinput/fire_emission_factors/fire_emission_factors_78PFTs_c20210803.nc'
  fire_emis_specifier = 'bc_a4 = 0.95 * BC',
                        'pom_a4 = 1.4 * OC',
			'num_a4 = num_a4_bc + num_a4_pom',
			'SO2 = 0.95 * SO2',
			'DMS = 0.95 * DMS',
			'SOAG = 0.45 * BIGALK + 0.35 * BIGENE + 8.2367 * TOLUENE + 1.4625 * BENZENE + 6.5013 * XYLENE + 0.34 * ISOP + 4.25 * C10H16'
/
```

**CMEPS: shr_fire_emis_mod.F90**
--------------------------------------
Both CAM and CLM read the above namelist in the routine `shr_fire_emis_mod.F90 `
in the directory `$SRCROOT/components/cmeps/cesm/nuopc_cap_share/`

```F90
    ! This reads the fire_emis_nl namelist group in drv_flds_in and parses the
    ! namelist information for the driver, CLM, and CAM.
    !
    ! Namelist variables:
    !   fire_emis_specifier, fire_emis_factors_file, fire_emis_elevated
    !
    !   fire_emis_specifier (array of strings) -- Each array element specifies
    !     how CAM-Chem constituents are mapped to basic smoke compounds in
    !     the fire emissions factors table (fire_emis_factors_file).  Each
    !     chemistry constituent name (left of '=' sign) is mapped to one or more
    !     smoke compound (separated by + sign if more than one), which can be
    !     proceeded by a multiplication factor (separated by '*').
    !     Example:
    !       fire_emis_specifier = 'bc_a1 = BC','pom_a1 = 1.4*OC','SO2 = SO2'
    !
    !   fire_emis_factors_file (string) -- Input file that contains the table
    !     of basic compounds that make up the smoke from the CLM fires.  This is
    !     used in CLM module FireEmisFactorsMod.
    !
    !   fire_emis_elevated (locical) -- If true then CAM-Chem treats the fire
    !     emission sources as 3-D vertically distributed forcings for the
    !     corresponding chemical tracers.
```

NOTE: the fire emis specifiers above are for CAM-chem and WILL NOT WORK with OSLO_AERO.
As an example, bc_a1 does not exist in OSLO_AERO but instead needs to be referenced as BC_A1. This is try for pom->OM, etc.

The following module data is then filled in from reading the namelist. 

```F90
  ! fire emissions component data structure (or user defined type)
  type shr_fire_emis_comp_t
     character(len=name_len)             :: name            ! emissions component name (in fire emissions input table)
     integer                             :: index
     real(r8), pointer                   :: emis_factors(:) ! function of plant-function-type (PFT)
     real(r8)                            :: molec_weight    ! molecular weight of the fire emissions compound (g/mole)
     type(shr_fire_emis_comp_t), pointer :: next_emiscomp   ! points to next member in the linked list
  endtype shr_fire_emis_comp_t
  type shr_fire_emis_comp_ptr
    type(shr_fire_emis_comp_t), pointer :: ptr ! points to fire emis data obj (shr_fire_emis_comp_t)
  endtype shr_fire_emis_comp_ptr

  ! chemical compound in CAM mechanism that has fire emissions
  type shr_fire_emis_mechcomp_t
     character(len=16)				:: name          ! compound name
     type(shr_fire_emis_comp_ptr), pointer	:: emis_comps(:) ! an array of pointers to fire emis components
     real(r8), pointer				:: coeffs(:)     ! coeffecients to emissions components
     integer					:: n_emis_comps  ! number of fire emis compounds that make up the emissions for this mechanis compound
  end type shr_fire_emis_mechcomp_t

  type(shr_fire_emis_mechcomp_t),  pointer :: shr_fire_emis_mechcomps(:) ! array of chemical compounds (in CAM mechanism) that have fire emissions
  type(shr_fire_emis_comp_t),      pointer :: shr_fire_emis_linkedlist   ! points to linked list top
```
**CLM: lnd2atmMod.F90**
----------------------------

The data sent to the mediator is computed in lnd2atmMod.F90 as follows:

```F90
     ! fire emissions fluxes
     if (shr_fire_emis_mechcomps_n>0) then
        call p2g(bounds, shr_fire_emis_mechcomps_n, &
            -fireemis_inst%fireflx_patch(bounds%begp:bounds%endp,:), &
             lnd2atm_inst%fireflx_grc   (bounds%begg:bounds%endg,:), &
             p2c_scale_type='unity', c2l_scale_type= 'unity', l2g_scale_type='unity')
        call p2g(bounds, &
             fireemis_inst%ztop_patch (bounds%begp:bounds%endp), &
             lnd2atm_inst%fireztop_grc(bounds%begg:bounds%endg), &
             p2c_scale_type='unity', c2l_scale_type= 'unity', l2g_scale_type='unity')
     endif
```

**CLM: CNFireEmissionsMod.F90**
---------------------------------------

The contents of fireems_inst%fireflx_patch are filled in the routine CNFireEmissionsMod.F90 (for non-FATES)

```F90
      real(r8) :: emis_flux(shr_fire_emis_comps_n)
      associate(fire_emis => fireemis_inst%fireflx_patch)

      do fp = 1,num_bgc_vegp
         p = filter_bgc_vegp(fp)
         g = patch%gridcell(p)
         c = patch%column(p)
         if (patch%itype(p) > 0)then

            ! loop over fire components
            emis_cmp => shr_fire_emis_linkedlist
            emis_cmp_loop: do while(associated(emis_cmp))
               icomp		   = emis_cmp%index
               epsilon		   = emis_cmp%emis_factors(patch%itype(p))
               comp(icomp)%emis(p) = epsilon * fire_flux* 1.e-3_r8/0.5_r8  ! (to convert gC/m2/sec to kg species/m2/sec)
               emis_flux(icomp)	   = comp(icomp)%emis(p)
               emis_cmp		   =>emis_cmp%next_emiscomp
            enddo emis_cmp_loop

            ! sum up the emissions compontent fluxes for the fluxes of chem mechanism compounds
            do imech = 1,shr_fire_emis_mechcomps_n
               n_emis_comps = shr_fire_emis_mechcomps(imech)%n_emis_comps
               do icomp = 1,n_emis_comps ! loop over number of emission components that make up the nth mechanism compoud
                  ii                 = shr_fire_emis_mechcomps(imech)%emis_comps(icomp)%ptr%index
                  fire_emis(p,imech) = fire_emis(p,imech) + shr_fire_emis_mechcomps(imech)%coeffs(icomp)*emis_flux(ii)
               enddo
               mech(imech)%emis(p) = fire_emis(p,imech)
            enddo
```

**Comment by mvertens** [comment]
@mvdebolskiy @kjetilaas @rosiealice @maritsandstad - this is the beginning of a description regarding how the data from CTSM fire emissions is parsed from the namelist and sent to the mediator and ingested by CAM. This is meant to address the question from  @mvdebolskiy regarding how CAM and CLM can coordinate indices based on what is sent by CTSM. Please add questions here.
I'll try to add more documentation before next week.

**Reply by rosiealice** [reply]
Thanks @mvertens! It seems that perhaps the major issue here is actually getting OsloAero to work with what exists already, given that FATES is really just going to be solving the same problem that CLM is solving - trying to provide the required fluxes in the right order? 

I have always been very scared of things in the shr directory, so thanks for demystifying it!

**Reply by mvdebolskiy** [reply]
Should the fields be standardized between CAM and OSLO-AERO or we are fiddling around with namelist building to convert the species names?

**Reply by mvertens** [reply]
We need to do 3 things - I think in the order below
1)  get FATES working with what exists already
2) get OsloAero to work with what exists already
3) get FATES working with Oslo-Aero

**Reply by mvertens** [reply]
@mvdebolskiy - we need to work with namelist building to get Oslo-AERO to work with the current fire-emissions. Neither oslo-aero or cam should need to change their naming conventions. I think there would be a lot of push back against that.

**Comment by maritsandstad** [comment]
@mvertens is Oslo-AERO working with current fire-emissions with non-FATES? Or does choosing it do a separate level of CAM to Oslo-AERO renaming from some namelist, or preset the fire namelist to fit it's naming conventions? Or is maybe this something we should do as part of step 2 and 3?

**Reply by mvertens** [reply]
No - Oslo-Aero is not working with current-fire-emissions with non-FATES. In fact if you look at the following discussion using fire emissions has not been used in previous production runs in CESM.
https://github.com/ESCOMP/CTSM/discussions/2397

**Reply by mvertens** [reply]
My feeling is that 3. will be very easy to do once 2. is done. The work for 1. will take the most time.

**Reply by maritsandstad** [reply]
Yeah, sounds like 1+2 should probably solve 3 by default more or less.

**Reply by mvertens** [reply]
Exactly.

**Comment by oyvindseland** [comment]
Hi
Does this set-up also  work for mct?

A suggestion for Oslo-Aero  included below.
I do not understand the 0.95 factors though, but have kept them for now.

&fire_emis_nl
  fire_emis_elevated = .false.
fire_emis_factors_file = '/glade/campaign/cgd/tss/lawrence/ccisensinput/fire_emission_factors/fire_emission_factors_78PFTs_c20210803.nc'
  fire_emis_specifier = 'BC_NI = 0.95*BC',
                        'OM_NI = 2.6 * OC',
			'SO2 = 0.975 * SO2',
			'SO4_PR = 0.025 * SO2',
			'DMS = 0.95 * DMS',
/

**Reply by mvertens** [reply]
@oyvindseland - I have no idea how this would work for mct and all of this is only based on the nuopc interface. I have no idea where the .95 came from - we should ask the CAM folks.

**Reply by mvertens** [reply]
@vitt - can you please comment on where the .95 factor comes from. That would be really appreciate.

**Reply by oyvindseland** [reply]
It does work for mct as in adding the fields, but does not give out any tendency terms

**Comment by DirkOlivie** [comment]
@oyvindseland, looks fine.  Possibly we can add:
'isoprene = ISOP',
'monoterp = C10H16'

For DMS, we maybe have to check that non-ocean emissions are allowed (will be the default in NorESM2.3).

I also don't know where the factor 0.95 comes from.

**Comment by oyvindseland** [comment]
@mvertens @gold2718 I am trying to test the file in the CMIP6 version of NorESM2. Can you download the file from glade. ?
BTW I added the definition in user_nl_cam, but this is then automatically included in drv_flds_in

**Reply by gold2718** [reply]
Where should this go, /cluster/shared/noresm/inputdata/lnd/clm2/firedata ?

**Reply by mvertens** [reply]
Derecho is down - and I think there is a new firedata dataset. Let me see where things stand once derecho is back up.

**Comment by oyvindseland** [comment]
The suggested path is fine

**Reply by gold2718** [reply]
Done, please let me know if there is a problem, otherwise, you can close this discussion.

**Reply by oyvindseland** [reply]
Included in namelist but all sigma computers have problems with storage now so not able to submit a test job

**Comment by oyvindseland** [comment]
Hi, The results as in change in concentrations looks reasonable when testing in the most recent version NorESM2-CMIP6 . However the fluxes has not changed except for round-off. E.g  organic carbon emissions is written out as SFOM_NI for surface emissions  and OM_NI_CMXF for 3D emissions. 
The fields has to be collected somewhere as a flux in the atmosphere somewhere since the concentrations are changing. Do you any of you know the name of the field so I can write it out?

**Comment by oyvindseland** [comment]
Link to a test simulation at Betzy Two months only but clear signal in concentrations
With fire emissions
Results: /cluster/work/users/oyvinds/archive/nf1850norbc.f19_f19_fireemis2.0.20240913
Case: /cluster/projects/nn2345k/oyvinds/NorESM2-CMIP6/cases/nf1850norbc.f19_f19_fireemis2.0.20240913

Base-line:
Results: /cluster/work/users/oyvinds/archive/nf1850norbc.f19_f19_default2.0.8.20240913
Case: /cluster/projects/nn2345k/oyvinds/NorESM2-CMIP6/cases/nf1850norbc.f19_f19_default2.0.8.20240913

**Comment by oyvindseland** [comment]
Technical coupler test Betzy 
/cluster/work/users/oyvinds/archive/nf1850norbc.f19_f19_fireemis2.0.20240919
and 
/cluster/projects/nn2345k/oyvinds/NorESM2-CMIP6/cases/nf1850norbc.f19_f19_fireemis2.0.20240919

**Comment by oyvindseland** [comment]
I  have tested the fire emissions in NorESM2.0.8 and it seems to work. I will make longer simulations using this set-up in user_nl_cam

&fire_emis_nl
fire_emis_elevated = .true.
fire_emis_factors_file = '/cluster/shared/noresm/inputdata/lnd/clm2/firedata/fir
e_emission_factors_78PFTs_c20210803.nc'
fire_emis_specifier = 'BC_NI = 1.0*BC',
'OM_NI = 2.6 * OC',
'SO2 = 0.975 * SO2',
'SO4_PR = 0.025 * SO2'
/

Modifying the driver namelist through the cam namelist is probably not the optimal way of doing it, but it looks like how the structure s set up.
The CAM fire emissions subroutine in NorESM2 is almost the same as the one in NorESM2.5 The  exception is how the coupled indices are set-up. The subroutine is  cam/src/chemistry/mozart/fire_emissions.F90. The fluxes are written out in the same subroutine but the fields have to be included either in the namelist or using add_default. Given the very similar structure it is likely that the same set-up will work also in 2.5

Given the settings above the fluxes are automatically added to the tracer fields as named. For the simulations I am planning to run I will try to compare it to the CMIP6 simulations so I will use the 3d vertical profile as defined in the model, and not include VOC and DMS from fires.  
At the moment it looks like the all emissions has to be included either as surface or 3D.
With respect to numbers: I am using 1 for BC not 0.95. I do not know the reasoning for the 0.95. May be numerical, may be related to elemental vs black carbon. NorESM2 is using a high OM/OC ratio of 2.6 (but from publications) so in principle yet another tuning factor. 
 0.025 * SO2 is an assumed value for rapid plume conversion to SO4 both in Oslo-Aero and trop-mam4
@MichaelSchulzMETNO

**Reply by oyvindseland** [reply]
NorESM2 simulation: /cluster/projects/nn2345k/oyvinds/NorESM2-CMIP6/cases/nf1850norbc_aer2014.f19_f19_progfirebb.20240926

**Comment by oyvindseland** [comment]
A similar setting with updated files also works for NorESM3 if it is set up with plain clm

Compset tested: 
```
1850_CAM70%LT%NORESM%CAMoslo_CLM60%BGC-CROP_CICE%PRES_DOCN%DOM_MOSART_DGLC%NOEVOLVE_SWAV_SESP
```

Resolution:
`ne16pg3_ne16pg3_mtn14
`

Add in user_nl_cam
```
&fire_emis_nl
fire_emis_elevated = .true.
fire_emis_factors_file = '/cluster/shared/noresm/inputdata/lnd/clm2/firedata/fire_emission_factors_78PFTs_c20240624.nc'
fire_emis_specifier = 'BC_NI = 1.0*BC',
'OM_NI = 2.6 * OC',
'SO2 = 0.975 * SO2',
'SO4_PR = 0.025 * SO2'
/
```

Paths on nird 
```
directory: 
   /cluster/projects/nn9560k/oyvinds/noresm3_0_beta02/cases/
cases:
nf1850_ne16pg3_ne16pg3_mtn14.noresm3_0_beta02_clmbiomas_20250912
nf1850_sstp4K_ne16pg3_ne16pg3_mtn14.noresm3_0_beta02_clmbiomas_20250912
nf2000_ne16pg3_ne16pg3_mtn14.noresm3_0_beta02_clmbiomtest.20250909
```

Not documented on noresm3-dev since it was not seen as relevant for the development
